### PR TITLE
Show all locations [WIP] [DO NOT CLOSE]

### PR DESCRIPTION
**Pull request by timabell**
**Note**: please keep this PR open as it's being used in user research.

### Context

Discussions with Paul H

### Changes proposed in this pull request

Show both provider and campus locations

* still need to sort out titles and grouping properly

## After

7 vm_357

### CODE REVIEW

**Comment by timabell** [comment]
[Update map to show *both* training location and provider location](https://trello.com/c/QkIb2KOT/481-update-map-to-show-both-training-location-and-provider-location)

**Comment by timabell** [comment]
Cheers for the feedback @defong we've done as you suggest and just dumped the data out to js. undefined etc needs front-end dev love, possibly in a different PR.

File: src/Assets/Javascript/map.js [hunk 1]
```diff
@@ -1,3 +1,26 @@
...
+
+// From here: https://www.movable-type.co.uk/scripts/latlong.html
+const distanceBetweenLatLng = (lat1, lng1, lat2, lng2) => {
+  var R = 6371e3 // Earth’s radius in metres.
+  var φ1 = toRadians(lat1)
+  var φ2 = toRadians(lat2)
+  var Δφ = toRadians(lat2 - lat1)
+  var Δλ = toRadians(lng2 - lng1)
+
+  var a = Math.sin(Δφ / 2) * Math.sin(Δφ / 2) + Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ / 2) * Math.sin(Δλ / 2)
```
Comment (defong) [inline src/Assets/Javascript/map.js]: are those maths symbol signs?

File: src/Assets/Javascript/map.js [hunk 2]
```diff
@@ -1,3 +1,26 @@
...
+  var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
+
+  var d = R * c
+  return d
+}
+
+distanceBetweenLatLng(50.854259, 0.573453, 51.5001754, -0.1332326)
```
Comment (defong) [inline src/Assets/Javascript/map.js]: is this meant to be there?

File: src/Assets/Javascript/map.js [hunk 3]
```diff
@@ -175,32 +209,171 @@ const initGoogleMaps = () => {
...
+    return postcode ? postcode.replace(" ", "") : null
+  }
+
+  // pins = {
+  //   'lat,lng': {
+  //       distance: meters,
+  //       locations: []
+  //   }
+  // }
+  const pins = locations.reduce((pins, location) => {
```
Comment (defong) [inline src/Assets/Javascript/map.js]: is this self invoke or something?

`const pins = locations.reduce((pins, location) => {`

I'm seeing pins as a variable and pins as a parameter

as well as 

`const pin = pins[key] || {}`

just sanity checking...

File: src/Assets/Javascript/map.js [hunk 4]
```diff
@@ -175,32 +209,171 @@ const initGoogleMaps = () => {
...
+  //   'lat,lng': {
+  //       distance: meters,
+  //       locations: []
+  //   }
+  // }
+
+  console.log("pins", pins)
+  console.log("sorted", Object.values(pins).sort((left, right) => left.distance - right.distance))
+  console.log("foundPostcodes", foundPostcodes)
+  console.log("foundLocationsWithoutPostcode", foundLocationsWithoutPostcode)
```
Comment (defong) [inline src/Assets/Javascript/map.js]: don't forget console.log

File: src/Assets/Javascript/map.js [hunk 5]
```diff
@@ -175,32 +209,171 @@ const initGoogleMaps = () => {
...
+            : `${Object.keys(pin.providers).length} providers`
+        }</span>
+        <span class="condensed-content">${
+          Object.keys(pin.providers).length === 1
+            ? Object.values(pin.providers)[0][0].url.split("/")[2]
+            : `${Object.keys(pin.providers).length} providers`
+        }</span>
+      `
+
+      const openContent = document.createElement("div")
```
Comment (defong) [inline src/Assets/Javascript/map.js]: can we break these into its own function
`closedContent`
`openContent`

File: src/Assets/Javascript/map.js [hunk 6]
```diff
@@ -252,6 +425,19 @@ const initGoogleMaps = () => {
...
     map.panTo(new google.maps.LatLng(centerLat, centerLng))
   }
+
+  const $courseLinks = document.querySelectorAll(".map-course-link")
+  Array.from($courseLinks).forEach($link => {
+    $link.addEventListener("click", evt => {
+      evt.preventDefault()
+      const href = evt.target.href
+      const popupLink = document.querySelector(`.popup-bubble-content [href="${"/course" + href.split("course")[1]}"]`)
+      // ⚠️ PROTOTYPE CODE QUARANTINE ZONE ️️⚠️
```
Comment (defong) [inline src/Assets/Javascript/map.js]: Love that bit

**Comment by timabell** [comment]
closing this as it's toooo old, open a new one when code is ready

**Comment by benilovj** [comment]
@timabell I've reopened this because we need the Heroku deployment for UR.

**Comment by timabell** [comment]
Could we open one without my name on it, it's giving me fourth-wall angst.

**Comment by defong** [comment]
orphaned as no trello card either close or follow up

**Pull request closed without merging**